src/LibraryManager/Providers/BaseProvider.cs

```diff
@@ -0,0 +1,297 @@
+            }
+
+            return LibraryOperationResult.FromSuccess(state);
+        }
+
+        private async Task<Stream> GetStreamAsync(ILibraryInstallationState state, string sourceFile, CancellationToken cancellationToken)
+        {
+            if (!string.IsNullOrEmpty(state.Name) && !string.IsNullOrEmpty(state.Version))
+            {
+                string absolute = Path.Combine(CacheFolder, state.Name, state.Version, sourceFile);
+
+                if (File.Exists(absolute))
+                {
+                    return await HostInteraction.ReadFileAsync(absolute, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return null;
+        }
+
+        private bool IsLibraryUpToDate(ILibraryInstallationState state)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(state.Name) && !string.IsNullOrEmpty(state.Version))
+                {
+                    string cacheDir = Path.Combine(CacheFolder, state.Name, state.Version);
+                    string destinationDir = Path.Combine(HostInteraction.WorkingDirectory, state.DestinationPath);
+
+                    foreach (string sourceFile in state.Files)
+                    {
+                        var destinationFile = new FileInfo(Path.Combine(destinationDir, sourceFile).Replace('\\', '/'));
+                        var cacheFile = new FileInfo(Path.Combine(cacheDir, sourceFile).Replace('\\', '/'));
+
+                        if (!destinationFile.Exists || !cacheFile.Exists || !FileHelpers.AreFilesUpToDate(destinationFile, cacheFile))
+                        {
+                            return false;
+                        }
+                    }
+                }
+            }
+            catch
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Copies ILibraryInstallationState files to cache
+        /// </summary>
+        /// <param name="state"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        private async Task<ILibraryOperationResult> RefreshCacheAsync(ILibraryInstallationState state, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return LibraryOperationResult.FromCancelled(state);
+            }
+
+            string libraryDir = Path.Combine(CacheFolder, state.Name, state.Version);
+
+            try
+            {
+                var librariesMetadata = new List<CacheFileMetadata>();
+                foreach (string sourceFile in state.Files)
+                {
+                    string cacheFile = Path.Combine(libraryDir, sourceFile);
+                    string url = GetDownloadUrl(state, sourceFile);
+
+                    var newEntry = new CacheFileMetadata(url, cacheFile);
+                    if (!librariesMetadata.Contains(newEntry))
```

Comment:
I don't think this Contains would every return, would it? CacheFileMetadata doesn't override Equals, and it's class - not a struct. So we are going to do reference equality checking here, right? Since it's a newly created object, that would never exist. Unless we override Equals().